<b>Exercise 1a)</b> <br/>
We have the expression $f(x) = \mathbf{a^Tx}$ where $a$ and $x$ are column vectors of length $n$. $f(x)$ is therefore a scalar.<br/>
$x$ has dimensions (n, 1)<br/>
The result (gradient) inherit transposed $x$ dimentions, (1, n)


<b>Exercise 1b)</b> <br/>
We write $\mathbf{a^Tx} = \sum_{j=0}^{n-1}a_{j}x_j$. The partial derivative with respect to $x_i$ is
$$\frac{\partial}{\partial x_i} \left(\sum_{j=0}^{n-1} a_jx_j\right) = a_i$$
Since only the $i = j$ term contributes, this holds for all components $i$, so the gradient is ($a_1, a_2, ..., a_n$) which is $\mathbf{a}^T$.

<b>Exercise 1c)</b> <br/>
$\mathbf{a^TAa} = \sum_{j=0}^{n-1} \sum_{i=0}^{n-1} a_jA_{ij}a_i$<br/>
For $i=k$: $A_{kj}a_j$ contributes<br/>
For $j=k$: $a_iA_{ik}$ contributes<br/>
Taking $\frac{\partial}{\partial a_k}$, we get<br/>
$$\frac{\partial}{\partial a_k}\left(\mathbf{a^TAa}\right) = \sum_{j=0}^{n-1}A_{kj}a_j + \sum_{i=0}^{n-1}a_iA_{ik}$$
The first sum is the $k$ th component of $(\mathbf{A}\boldsymbol{a})$, and the second sum is the $k$ th component of $(\boldsymbol{a}^T \mathbf{A})^T = (\mathbf{A}^T \boldsymbol{a})$.
$$\left(\mathbf{Aa}\right)^T + \left(\mathbf{A}^T\mathbf{a}\right)^T = \mathbf{a}^T\mathbf{A}^T + \mathbf{a}^T\mathbf{A} = \mathbf{a}^T\left(\mathbf{A}+\mathbf{A}^T\right)$$

<b>Exercise 2a)</b><br/>
Taking the derivative of the error with respect to $\boldsymbol{\theta}$ and setting it to zero finds the minimum because we are dealing with a convex quadratic cost function. The squared error $||\mathbf{y} - \mathbf{X}\boldsymbol{\theta}||^2$ has a single global minimum. Setting the gradient to zero gives the critical point, which for a convex function is the global minimum. In short, solving $\nabla_{\theta} ||y - X\theta||^2 = 0$ gives the $\boldsymbol{\theta}$ that minimizes the error.

<b>Exercise 2b)</b><br/>
If $\mathbf{X}$ is square and invertible, the optimal solution is the one that gives zero error by exactly solving $\mathbf{X}\boldsymbol{\theta} = \mathbf{y}$. In that case, we can multiply both sides by $\mathbf{X}^{-1}$ to get
$$\boldsymbol{\theta} = \mathbf{X}^{-1}\mathbf{y}$$

<b>Exercise 2c)</b><br/>
$$f(s) = \left( \mathbf{x}-\mathbf{As}\right)^T\left(\mathbf{x}-\mathbf{As}\right)$$
We differentiate $f(s)$ with respect to $\boldsymbol{s}$. This expression is the squared norm of the error vector $\boldsymbol{e} = \boldsymbol{x} - \mathbf{A}\boldsymbol{s}$.<br/>
$$\frac{\partial}{\partial s}\left(\mathbf{e}^T\mathbf{e}\right) = 2\mathbf{e}^T\frac{\partial e}{\partial s}$$
Here $\boldsymbol{e} = \boldsymbol{x} - \mathbf{A}\boldsymbol{s}$, so $\frac{\partial \boldsymbol{e}}{\partial \boldsymbol{s}} = -\mathbf{A}$
$$\frac{\partial}{\partial s}\left(\mathbf{e}^T\mathbf{e}\right) = 2\mathbf{e}^T\left(-\mathbf{A}\right) = -2\left(\mathbf{x} - \mathbf{As}\right)^T\mathbf{A}$$

<b>Exercise 2d)</b><br/>
In OLS, the error vector is $\boldsymbol{y} - \mathbf{X}\boldsymbol{\theta}$. Using the formula above, and substituting $\mathbf{x}= \mathbf{y}$, $\mathbf{A}= \mathbf{X}$ and $\mathbf{s}= \mathbf{\theta}$, we get
$$-2\left(\mathbf{y} - \mathbf{X\theta}\right)^T\mathbf{X}.$$
Optimizing this gives
$$\left(\mathbf{y} - \mathbf{X\theta}\right)^T\mathbf{X} = 0^T$$
$$\mathbf{X}^T\left(\mathbf{y} - \mathbf{X\theta}\right) = 0$$
$$\mathbf{X}^T\mathbf{y} - \mathbf{X}^T\mathbf{X\theta} = 0$$
$$\mathbf{X}^T\mathbf{y} = \mathbf{X}^T\mathbf{X\theta}$$
Assuming $\mathbf{X}^T \mathbf{X}$ is invertible, we multiply by $(\mathbf{X}^T\mathbf{X})^{-1}$ to get
$$\theta = (\mathbf{X}^T\mathbf{X})^{-1}\mathbf{X}^T\mathbf{y}$$

<b>Exercise 3a)</b><br/>

In [40]:
import numpy as np

In [41]:
n = 20
income = np.array([116., 161., 167., 118., 172., 163., 179., 173., 162., 
                   116., 101., 176., 178., 172., 143., 135., 160., 101., 149., 125.])
children = np.array([5, 3, 0, 4, 5, 3, 0, 4, 4, 3, 3, 5, 1, 0, 2, 3, 2, 1, 5, 4])
spending = np.array([152., 141., 102., 136., 161., 129.,  99., 159., 160., 
                     107.,  98., 164., 121.,  93., 112., 127., 117.,  69., 156., 131.])

In [42]:
X = np.zeros((n, 3))
X[:, 0] = 1
X[:, 1] = income
X[:, 2] = children

In [45]:
print("Controlling the feature matrix X:")
print(X[:5])
print("Shape of X:", X.shape)

Controlling the feature matrix X:
[[  1. 116.   5.]
 [  1. 161.   3.]
 [  1. 167.   0.]
 [  1. 118.   4.]
 [  1. 172.   5.]]
Shape of X: (20, 3)


<b>Exercise 3b)</b><br/>

In [48]:
def OLS_parameters(X, y):
    return np.linalg.inv(X.T @ X) @ (X.T @ y)
beta = OLS_parameters(X, spending)
print("Optimal OLS parameters beta:", beta)

Optimal OLS parameters beta: [ 9.12808583  0.5119025  14.60743095]


The model would then be </br>
spending $\approx$ 9.13 + 0.512 $\cdot$ income + 14.61 $\cdot$ children

<b>Exercise 4a)</b><br/>